rclcpp/include/rclcpp/topic_statistics/subscriber_topic_statistics.hpp

```diff
@@ -0,0 +1,194 @@
+    }
+
+    subscriber_statistics_collectors_.clear();
+
+    if (publisher_timer_) {
+      publisher_timer_->cancel();
+      publisher_timer_.reset();
+    }
+    if (publisher_) {
+      publisher_.reset();
+    }
+  }
+
+  /// Set the timer used to publish statistics messages.
+  /**
+   * @param measurement_timer the timer to fire the publisher, created by the node
+   */
+  void SetPublisherTimer(const rclcpp::TimerBase::SharedPtr & publisher_timer)
+  {
+    publisher_timer_ = std::move(publisher_timer);
+  }
+
+  /// Handle a message received by the subscription to collect statistics.
+  /**
+   * @param received_message the message received by the subscription
+   * @param now_nanoseconds current time in nanoseconds
+   */
+  virtual void OnMessageReceived(
+    const CallbackMessageT & received_message,
+    const rcl_time_point_value_t now_nanoseconds) const
+  {
+    (void) received_message;
+
+    for (auto & collector : subscriber_statistics_collectors_) {
+      collector->OnMessageReceived(received_message, now_nanoseconds);
+    }
+  }
+
+  /// Return a vector of all the currently collected data
+  /**
+   * @return a vector of all the collected data
+   */
+  std::vector<StatisticData> GetCurrentCollectorData() const
+  {
+    std::vector<StatisticData> data;
+    for (const auto & collector : subscriber_statistics_collectors_) {
+      data.push_back(collector->GetStatisticsResults());
+    }
+    return data;
+  }
+
+private:
+  /// Publish a populated MetricsStatisticsMessage
+  virtual void PublishMessage()
+  {
+    rclcpp::Time window_end{GetCurrentNanosecondsSinceEpoch()};
+
+    for (auto & collector : subscriber_statistics_collectors_) {
+      const auto collected_stats = collector->GetStatisticsResults();
+
+      auto message = libstatistics_collector::collector::GenerateStatisticMessage(
+        node_name_,
+        collector->GetMetricName(),
+        collector->GetMetricUnit(),
+        window_start_,
+        window_end,
+        collected_stats);
+      publisher_->publish(message);
+    }
+    window_start_ = window_end;
+  }
+
+  /// Return the current nanoseconds (count) since epoch
+  /**
+   * Based on design discussions, using harccoded time instead of a node's clock
+   * due to lifecycle issues.
+   * @return the current nanoseconds (count) since epoch
+   */
+  int64_t GetCurrentNanosecondsSinceEpoch()
+  {
+    const auto now = std::chrono::system_clock::now();
+    return std::chrono::duration_cast<std::chrono::nanoseconds>(now.time_since_epoch()).count();
+  }
+
+  /// Collection of statistics collectors
+  std::vector<std::unique_ptr<topic_stats_collector>> subscriber_statistics_collectors_{};
+  /// Node name used to generate topic statistics messages to be published
+  const std::string node_name_;
+  /// Publisher, created by the node, used to publish topic statistics messages
+  rclcpp::Publisher<metrics_statistics_msgs::msg::MetricsMessage>::SharedPtr publisher_{nullptr};
```

Comment:
Why SharedPtr?

Reply:
https://github.com/ros2/rclcpp/blob/0f0b83368a9650b76152f017c21d51f4366245fd/rclcpp/include/rclcpp/publisher_base.hpp#L55 It uses shared from this, and I assume this class owns it? If not I guess it could be a `weak_ptr` instead.

Reply:
The `SubscriberTopicsStatistics` class owns the publisher. Please see the new constructor, that should hopefully clarify this, [where the publisher is created](https://github.com/ros2/rclcpp/pull/1050/files#diff-756335e5efe4ced64021eaf3abf4df29R97).